data race in `storage.TestRangeDataIteratorEmptyRange()`

https://circleci.com/gh/tschottdorf/cockroach/275

```
W0120 21:36:32.381313     796 id_alloc.go:85] allocator key is currently set at -1014; minID is 2; allocating again to skip 1016 IDs
I0120 21:36:32.387953     796 queue.go:124] processing range 2 from test queue with priority 2.000000...
I0120 21:36:32.388153     796 queue.go:124] processing range 1 from test queue with priority 1.000000...
I0120 21:36:32.388244     796 queue.go:124] processing range 1 from test queue with priority 3.000000...
I0120 21:36:32.388348     796 queue.go:124] processing range 2 from test queue with priority 2.000000...
I0120 21:36:32.388463     796 queue.go:124] processing range 1 from test queue with priority 3.000000...
2015/01/20 21:36:32 method Validate has wrong number of ins: 1
2015/01/20 21:36:32 raft: 1 became follower at term 5
2015/01/20 21:36:32 raft: newRaft 1 [peers: [1], term: 5, commit: 10, lastindex: 10, lastterm: 5]
==================
WARNING: DATA RACE
Write by goroutine 18:
  github.com/cockroachdb/cockroach/storage.TestRangeDataIteratorEmptyRange()
      /go/src/github.com/cockroachdb/cockroach/storage/range_data_iter_test.go:79 +0x1f8
  testing.tRunner()
      /usr/src/go/src/testing/testing.go:447 +0x133

Previous read by goroutine 100:
  github.com/cockroachdb/cockroach/storage.func·016()
      /go/src/github.com/cockroachdb/cockroach/storage/store.go:401 +0x10c
  sort.Search()
      /usr/src/go/src/sort/search.go:66 +0x66
  github.com/cockroachdb/cockroach/storage.(*Store).maybeSplitRangesByConfigs()
      /go/src/github.com/cockroachdb/cockroach/storage/store.go:402 +0x19f
  github.com/cockroachdb/cockroach/storage.(*Store).configGossipUpdate()
      /go/src/github.com/cockroachdb/cockroach/storage/store.go:386 +0x48b
  github.com/cockroachdb/cockroach/storage.*Store.(github.com/cockroachdb/cockroach/storage.configGossipUpdate)·fm()
      /go/src/github.com/cockroachdb/cockroach/storage/store.go:355 +0x51
  github.com/cockroachdb/cockroach/gossip.func·007()
      /go/src/github.com/cockroachdb/cockroach/gossip/infostore.go:274 +0x111

Goroutine 18 (running) created at:
  testing.RunTests()
      /usr/src/go/src/testing/testing.go:555 +0xd4e
  testing.(*M).Run()
      /usr/src/go/src/testing/testing.go:485 +0xe0
  main.main()
      github.com/cockroachdb/cockroach/storage/_test/_testmain.go:262 +0x28c

Goroutine 100 (finished) created at:
  github.com/cockroachdb/cockroach/gossip.(*infoStore).processCallbacks()
      /go/src/github.com/cockroachdb/cockroach/gossip/infostore.go:276 +0x40a
  github.com/cockroachdb/cockroach/gossip.(*infoStore).addInfo()
      /go/src/github.com/cockroachdb/cockroach/gossip/infostore.go:225 +0x63f
  github.com/cockroachdb/cockroach/gossip.(*Gossip).AddInfo()
      /go/src/github.com/cockroachdb/cockroach/gossip/gossip.go:162 +0x187
  github.com/cockroachdb/cockroach/storage.(*Range).maybeGossipConfigs()
      /go/src/github.com/cockroachdb/cockroach/storage/range.go:583 +0x69f
  github.com/cockroachdb/cockroach/storage.(*Range).start()
      /go/src/github.com/cockroachdb/cockroach/storage/range.go:203 +0x85
  github.com/cockroachdb/cockroach/storage.(*Store).addRangeInternal()
      /go/src/github.com/cockroachdb/cockroach/storage/store.go:656 +0x5b
  github.com/cockroachdb/cockroach/storage.(*Store).AddRange()
      /go/src/github.com/cockroachdb/cockroach/storage/store.go:638 +0x77
  github.com/cockroachdb/cockroach/storage.(*testContext).Start()
      /go/src/github.com/cockroachdb/cockroach/storage/range_test.go:117 +0xcbd
  github.com/cockroachdb/cockroach/storage.TestRangeDataIteratorEmptyRange()
      /go/src/github.com/cockroachdb/cockroach/storage/range_data_iter_test.go:70 +0xaf
  testing.tRunner()
      /usr/src/go/src/testing/testing.go:447 +0x133
==================
```

Plus this one could mean something:
```
--- FAIL: TestScannerTiming (0.02s)
	scanner_test.go:203: expected 2.5ms, got 6.377837ms: exceeds max error of 1.5ms
```

--- a/storage/range.go
+++ b/storage/range.go
@@ -240,9 +240,9 @@ func (r *Range) GetReplica() *proto.Replica {
 }
 
 // ContainsKey returns whether this range contains the specified key.
+// Read-lock the mutex to protect access to Desc, which might be changed
+// concurrently via range split.
 func (r *Range) ContainsKey(key proto.Key) bool {
-	// Read-lock the mutex to protect access to Desc, which might be changed
-	// concurrently via range split.
 	r.RLock()
 	defer r.RUnlock()
 	return r.Desc.ContainsKey(engine.KeyAddress(key))

--- a/storage/range_data_iter_test.go
+++ b/storage/range_data_iter_test.go
@@ -74,9 +74,11 @@ func TestRangeDataIteratorEmptyRange(t *testing.T) {
 	// records and config entries during the iteration. This is a rather
 	// nasty little hack, but since it's test code, meh.
 	tc.rng.Lock()
+	tc.store.mu.Lock()
 	newDesc := *tc.rng.Desc
 	newDesc.StartKey = proto.Key("a")
 	tc.rng.Desc = &newDesc
+	tc.store.mu.Unlock()
 	tc.rng.Unlock()
 
 	iter := newRangeDataIterator(tc.rng, tc.rng.rm.Engine())
@@ -98,10 +100,12 @@ func TestRangeDataIterator(t *testing.T) {
 
 	// See notes in EmptyRange test method for adjustment to descriptor.
 	tc.rng.Lock()
+	tc.store.mu.Lock()
 	newDesc := *tc.rng.Desc
 	newDesc.StartKey = proto.Key("b")
 	newDesc.EndKey = proto.Key("c")
 	tc.rng.Desc = &newDesc
+	tc.store.mu.Unlock()
 	tc.rng.Unlock()
 
 	// Create two more ranges, one before the test range and one after.

--- a/storage/store.go
+++ b/storage/store.go
@@ -406,20 +406,21 @@ func (s *Store) maybeSplitRangesByConfigs(configMap PrefixConfigMap) {
 			s.mu.Unlock()
 			continue
 		}
-		desc := *s.rangesByKey[n].Desc
+		start := s.rangesByKey[n].Desc.StartKey
+		end := s.rangesByKey[n].Desc.EndKey
 		s.mu.Unlock()
 
 		// Now split the range into pieces by intersecting it with the
 		// boundaries of the config map.
-		splits, err := configMap.SplitRangeByPrefixes(desc.StartKey, desc.EndKey)
+		splits, err := configMap.SplitRangeByPrefixes(start, end)
 		if err != nil {
-			log.Errorf("unable to split range %q-%q by prefix map %s", desc.StartKey, desc.EndKey, configMap)
+			log.Errorf("unable to split range %q-%q by prefix map %s", start, end, configMap)
 			continue
 		}
 		// Gather new splits.
 		var splitKeys []proto.Key
 		for _, split := range splits {
-			if split.end.Less(desc.EndKey) {
+			if split.end.Less(end) {
 				splitKeys = append(splitKeys, split.end)
 			}
 		}
@@ -427,7 +428,7 @@ func (s *Store) maybeSplitRangesByConfigs(configMap PrefixConfigMap) {
 			continue
 		}
 		// Invoke admin split for each proposed split key.
-		log.Infof("splitting range %q-%q at keys %v", desc.StartKey, desc.EndKey, splitKeys)
+		log.Infof("splitting range %q-%q at keys %v", start, end, splitKeys)
 		for _, splitKey := range splitKeys {
 			req := &proto.AdminSplitRequest{
 				RequestHeader: proto.RequestHeader{Key: splitKey},

